### DISCUSSION TITLE
Uses of mjolnir.tile_extract in Valhalla 3.1.2

### DISCUSSION THREAD

**Discussion by mloskot** [valhalla/valhalla#3201] [Q&A] [answered]
What are the uses of the `tile_extract`?
Does `valhalla_build_tiles` generate the `.tar` with the tile extract?
Does `valhalla_run_route` expect to load the extract, along with the `tile_dir`?
It does issue an error if the extract can not be loaded:
```
2021/07/08 16:32:16.843964 [ERROR] (stat): /data/valhalla/tiles.tar No such file or directory
2021/07/08 16:32:16.844405 [WARN] Tile extract could not be loaded
```
How the extract is supposed to be generated nowadays?

Since PR https://github.com/valhalla/valhalla/pull/2734 the `build_tile_set` function explicitly deactivates `tile_extract` configuration option as unused:

https://github.com/valhalla/valhalla/blob/d7a4e787615c54e2bf72fd452ba47d59051005c1/src/mjolnir/util.cc#L178-L180

### References

- https://github.com/valhalla/docker/issues/34

**Comment by nilsnolde** [comment, marked as answer]
Can't remember exactly but I think the tarred tiles perform better when reading for some fancy CS reason;)

It's not _really_ required, it's enough to have tiles in `tile_dir` but not have a tar file, for ~~all~~ (well, most..) executables (it's the GraphReader that sets up the tile cache etc). That error you're seeing should really be a warning IMO.

> Does valhalla_build_tiles generate the .tar with the tile extract?

Nope you need to do that, e.g. `find site/valhalla_tiles | sort -n | tar cf "./site/valhalla_tiles.tar" --no-recursion -T -`.

**Comment by nilsnolde** [comment]
BTW, do you know https://github.com/valhalla/valhalla#valhalla_service-aka-one-shot-mode? Unless you really want valhalla_run_route.. only (to me) known reason to run that is for pseudo-benchmarking

**Reply by mloskot** [reply]
Yes, I've learned about it.
I'm just trying my old workflows with build of newer Valhalla and I found command line interfaces of `valhalla_run_route` changed and it made me wonder where else things changed that may be affecting my (old) workflows.

**Comment by kevinkreiser** [comment]
this config option is only used in the running service, @mloskot we deliberately remove that when doing data builds because we need to modify the tiles and we dont support doing that to them when they are in a tar file. essentially we ignore the tile extract in the config whenever doing tilebuilding stuff (ie mjolnir)

> some fancy CS reason

hahaha. yes when running any of the service type endpoints (especially a long running service that will be used to handle many requests) the tar of tiles uses memory-mapping which has performance and resource allocation benefits over a directory of files. i should also mention that that warning is really just to show you that you configured the service to look for a tar that wasnt there. it always checks that location before falling back to `tile_dir`

**Reply by mloskot** [reply]
> we deliberately remove that when doing data builds because we need to modify the tiles and we dont support doing that to them when they are in a tar file

Yes, I figured that limitation.

> warning is really just to show you that you configured the service to look for a tar that wasnt there

Looks like I misinterpreted the `[ERROR] (stat): /data/valhalla/tiles.tar No such file or directory` as the option is mandatory,
whereas optional properties are enabled/disabled via their presence/absence from the config file.

**Comment by mloskot** [comment]
Thank you both for prompt clarification.
